### ISSUE TITLE
Serializer exception when using Phobos and Akka.Hosting

### ISSUE THREAD

**Issue by gerardpeters** [petabridge/phobos-issues#57]
Hi,

I changed parts of our system to use the new Akka.Hosting with Phobos. This works well until I add a ShardRegion to the service. In Lighthouse an exception is logged:

```
System.Runtime.Serialization.SerializationException: Cannot find serializer with id [210] (manifest [WTRACE]). The most probable reason is that the configuration entry 'akka.actor.serializers' is not in sync between the two systems. Serializer Id [210] is not one of the internal Akka.NET serializer. Please contact your plugin provider for more information.
   at Akka.Serialization.Serialization.Deserialize(Byte[] bytes, Int32 serializerId, String manifest)
   at Akka.Remote.Serialization.WrappedPayloadSupport.PayloadFrom(Payload payload)
   at Akka.Remote.Serialization.MessageContainerSerializer.FromBinary(Byte[] bytes, Type type)
   --- End of inner exception stack trace ---
   at Akka.Remote.Serialization.MessageContainerSerializer.FromBinary(Byte[] bytes, Type type)
   at Akka.Serialization.Serialization.Deserialize(Byte[] bytes, Int32 serializerId, String manifest)
   at Akka.Remote.MessageSerializer.Deserialize(ExtendedActorSystem system, Payload messageProtocol)
   at Akka.Remote.DefaultMessageDispatcher.Dispatch(IInternalActorRef recipient, Address recipientAddress, Payload message, IActorRef senderOption)
   at Akka.Remote.EndpointReader.<Reading>b__11_0(InboundPayload inbound)
```

Found that serializer 210 is used by Phobos. When removing the WithShardRegion from the startup everything is fine.
The code I use in the startup of the service with the sharding region:
```
services.AddAkka(DireXtion.Akka.Shared.Constants.ActorSystemName, configurationBuilder =>
    {
        var baseConfigPath = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)!, "akka.conf");
        configurationBuilder
            .AddHocon(ConfigurationFactory.ParseString(File.ReadAllText(baseConfigPath)))
            .WithRemoting(akkaConfiguration.Host, akkaConfiguration.Port)
            .WithClustering(new ClusterOptions
            {
                SeedNodes = akkaConfiguration.SeedNodes.Select(Akka.Actor.Address.Parse).ToArray(),
            })
            .WithShardRegionProxy<ParcelProxy>(ParcelMessageExtractor.ClusterShardingName, ParcelMessageExtractor.ClusterShardingRole, new ParcelMessageExtractor())
            .WithPhobos(AkkaRunMode.AkkaCluster, new PhobosNoiseTraceFilter())
            .WithActors((system, registry) =>
            {
                var dr = DependencyResolver.For(system);
                var masterActor = system.ActorOf(dr.Props<MasterActor>(), "master");
                var orderServiceCoordinator = system.ActorOf(Props.Empty.WithRouter(FromConfig.Instance), "create-visit-group");

                registry.TryRegister<MasterActor>(masterActor);
                registry.TryRegister<OrderServiceCoordinator>(orderServiceCoordinator);
            });
    });
```

And the hocon inside akka.conf
```
petabridge.cmd{
	# default IP address used to listen for incoming petabridge.cmd client connections
	# should be a safe default as it listens on "all network interfaces".
	host = "0.0.0.0"

	# default port number used to listen for incoming petabridge.cmd client connections
	port = 9211
}

akka {
    loggers = ["Akka.Logger.Serilog.SerilogLogger, Akka.Logger.Serilog"]
	actor {
		deployment {                  
			"/master/pre-order-entry" {
                router = consistent-hashing-pool
                nr-of-instances = 12
                cluster {
				    enabled = on
					allow-local-routees = on
					use-role = order-entry-service
                    max-nr-of-instances-per-node = 3
				}
            }
            "/master/order-entry" {
                router = consistent-hashing-pool
                nr-of-instances = 12
                cluster {
				    enabled = on
					allow-local-routees = on
					use-role = order-entry-service
                    max-nr-of-instances-per-node = 3
				}
            }
            "/create-visit-group" {
                router = consistent-hashing-group
                routees.paths = ["/user/coordinator"]
				nr-of-instances = 12
				cluster {
				    enabled = on										  
					allow-local-routees = on
					use-role = order-service
                    max-nr-of-instances-per-node = 3
				}
            }
		}
	}
	
	cluster {
		roles = ["order-entry-service"]
        downing-provider-class = "Akka.Cluster.SplitBrainResolver, Akka.Cluster"
	}
}
```

Am I missing something here? We use Phobos in another project with ShardRegions without akka.hosting and I can't see any errors there.

**Comment by Aaronontheweb** [comment]
@gerardpeters two possibilities here:

1. Lighthouse shares the same `akka.cluster.role` as the `ParcelMessageExtractor.ClusterShardingRole` and is thus receiving some of the sharding messages which get traced by Phobos or
2. Akka.Hosting is injecting the configuration it produces from this fluent builder incorrectly, which is causing Lighthouse to pick up messages from the sharding system that it should not be receiving.

In either case this probably isn't a Phobos issue - it looks like a configuration issue with the sharding system. I'll see if I can reproduce this in Akka.Hosting.

**Comment by Aaronontheweb** [comment]
I've checked the Akka.Hosting code and can confirm this is not an issue in the library there - we use that `ShardRegionProxy` line in a couple of areas inside our own demo applications and it works as expected. Can you check to make sure you're not passing in a `null` or `string.Empty` value for the role name in your `WithShardRegionProxy` call? Or that the role name doesn't equal one used by lighthouse?

**Comment by gerardpeters** [comment]
@Aaronontheweb 
I logged the config on start. Lighthouse has one role, which is Lighthouse. The RoleName for the shardRegion is a constant which is indeed not `null` or `string.empty`. Removing the `WithPhobos` line removes the exception.
I looked at the sql sharding example and didn't see anything obvious that I'm missing

**Comment by Aaronontheweb** [comment]
Can you share a copy of your Lighthouse setup? The issue is the sharding messages going where they should not - the trace serialization errors are just highlighting that

**Comment by gerardpeters** [comment]
I was using an old version of the the lighthouse repository. I tried the docker image
```docker run --name lighthouse1 --hostname lighthouse1 -p 9153:9153 -p 9110:9110 --env ACTORSYSTEM=dirextion-actors --env CLUSTER_IP=lighthouse1 --env CLUSTER_PORT=9153 --env CLUSTER_SEEDS="akka.tcp://dirextion-actors@lighthouse1:9153" petabridge/lighthouse:latest
```
But without any success.. 

```
[WARNING][05/12/2022 07:34:51][Thread 0021][akka.tcp://dirextion-actors@lighthouse1:9153/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2Fdirextion-actors%40127.0.0.1%3A50109-2/endpointWriter/endpointReader-akka.tcp%3A%2F%2Fdirextion-actors%40127.0.0.1%3A50109-1] Serializer not defined for message with serializer id [6] and manifest []. Transient association error (association remains live). Failed to deserialize payload object when deserializing ActorSelectionMessage addressed to [system,cluster,heartbeatReceiver]
```

**Comment by Aaronontheweb** [comment]
I'll see if I can reproduce this @gerardpeters

**Comment by Arkatufus** [comment]
@gerardpeters 
Which lighthouse version did you use? I used the latest in the sharding sqlserver sample project and it works just fine.
This is the `docker-compose.yaml` file that I used to set up the project:
```yaml
version: '3.7'

services:
  sqlsharding-sql:
    image: akkadotnet.sqlserver:0.1.0
    ports:
      - "1533:1433"
    environment:
      ACCEPT_EULA: "Y"
      SA_PASSWORD: "yourStrong(!)Password"
  
  zipkin:
    image: openzipkin/zipkin:latest
    hostname: zipkin
    container_name: zipkin
    restart: always
    ports:
      - "9411:9411"
    
  lighthouse:
    image: petabridge/lighthouse:latest
    ports:
      - '9110:9110'
      - '4053:4053'
    environment:
      ACTORSYSTEM: "SqlSharding"
      CLUSTER_PORT: 4053
      CLUSTER_IP: "localhost"
      CLUSTER_SEEDS: "akka.tcp://SqlSharding@localhost:4053"
```

**Comment by Arkatufus** [comment]
The code to setup Akka in the web app program is quite similar to your code snippet:
```c#
builder.Services.AddAkka("SqlSharding", (configurationBuilder, provider) =>
{
    configurationBuilder
        .WithRemoting(hostName, port)
        .WithCustomSerializer("sql-sharding", new[] { typeof(ISqlShardingProtocolMember) }, system => new MessageSerializer(system))
        .WithClustering(new ClusterOptions { Roles = new[] { "Web" }, SeedNodes = seeds })
        .WithShardRegionProxy<ProductMarker>("products", ProductActorProps.SingletonActorRole, new ProductMessageRouter())
        .WithPhobos(AkkaRunMode.AkkaCluster)
        .WithActors((system, registry) =>
        {
            var proxyProps = system.ProductIndexProxyProps();
            registry.TryRegister<ProductIndexMarker>(system.ActorOf(proxyProps, "product-proxy"));
        });
});
```

**Comment by gerardpeters** [comment]
I used the latest 1.6.1 version. But it seems that the problem is when using the overload for `WithPhobos` to register an `ITraceFilter`. 
When i use `.WithPhobos(AkkaRunMode.AkkaCluster)` I don't get any errors in Lighthouse. But when I use `.WithPhobos(AkkaRunMode.AkkaCluster, new PhobosNoiseTraceFilter())` the deserialization errors are thrown and the cluster gets unstable.

The `PhobosNoiceTraceFilter` snippet:
```
public class PhobosNoiseTraceFilter : ITraceFilter
{
    public bool ShouldTraceMessage(object message, bool alreadyInTrace)
    {
        switch (message)
        {
            case global::Akka.IO.Tcp.Event _:
            case global::Akka.Cluster.ClusterEvent.MemberStatusChange _:
            case global::Akka.DistributedData.FlushChanges _:
                return false;
        }

        var type = message.GetType();
        return type.Namespace switch
        {
            "Akka.Actor.Scheduler" => false,
            "Akka.DistributedData.Internal" => false,
            "Akka.Cluster.Sharding" => false,
            _ => alreadyInTrace,
        };
    }
}
```
Without this tracefilter the traces are getting very noisy.

**Comment by Arkatufus** [comment]
@gerardpeters Thank you for the last hint, I finally able to reproduce the bug. We'll work on this ASAP

**Comment by gerardpeters** [comment]
Hi @Arkatufus, do you have any ETA? And is this problem specific to the filter we use?

**Comment by Aaronontheweb** [comment]
I've been able to reproduce this issue, and it was _a lot_ trickier than I anticipated. However, I believe this is now fixed and we should have a release soon.

**Comment by Aaronontheweb** [comment]
I have a fix for this available but I need to return to my home computer and run NCrunch on it for a while (out of the office today) before I sign off on it - the fix is no good if it only solves the problem 99.9% of the time. Release probably won't happen until Friday at this rate.

**Comment by Aaronontheweb** [comment]
Fix is now available in Phobos 2.0.3 https://phobos.petabridge.com/articles/releases/RELEASE_NOTES.html#203httpssdkbincompublisherpetabridgeproductphobospackagesphobosactorversions203-may-26th-2022
